### PR TITLE
OCPBUGS-77344: enable dcm on route changes

### PR DESCRIPTION
**Pull request by jcmoraisjr**
DCM was disabled on route changes due to the way the resource changes are being handled: without DCM, the template plugin applies a route remove, followed by a route add, which works well in the fork-and-reload approach since the removal is not applied until adding the route again succeed. On DCM, each of these operations are done separately as well, which means that the route is completely disabled during some time.

This update changes this behavior by saving previous and updated states of a route, identifying the differences, and dynamically applying them if possible. A fork-and-reload is made in case the dynamic update is not possible, as before.

Part of the changes also include extending the Endpoint struct so that it stores the weight and the VerifyHostname properties of each endpoint in order to make comparisons easier, as well as decrease the number of parameters sent to inner methods. These fields cannot be updated during the Endpoint creation, since they depend on the backend the endpoint is assigned. The router's updateEndpointTable() method centralizes all changes to these context based fields.

Also, reverting https://github.com/openshift/router/pull/739 as a separate commit since it fixes the root cause of the problem.

https://redhat.atlassian.net/browse/OCPBUGS-77344

## Summary by CodeRabbit

* **New Features**
  * HAProxy error-detection helpers for common server conditions.
  * Updates now report when an endpoint was replaced vs updated in-place.

* **Bug Fixes**
  * Disabling health checks explicitly marks servers healthy and recovers when checks are missing.
  * Improved handling of protocol/port/weight changes to avoid misconfiguration and failures.

* **Refactor**
  * Centralized endpoint table management and streamlined add/update/delete flows for less disruptive dynamic route updates.

* **Tests**
  * Expanded tests covering update/replace scenarios and HTTP/2 cleartext edge cases.

### CODE REVIEW

**Comment by Miciah** [comment]
It would be helpful cross-reference #739 in the PR description and commit 0e47c836e9f2e5bbbf95c35604fc89a6fa8b32ee in the commit message—or even have a separate revert commit if it reverts cleanly.

File: pkg/router/template/configmanager/haproxy/backend.go [hunk 1]
```diff
@@ -331,33 +343,63 @@ func (b *Backend) AddServer(cfg *templaterouter.ServiceAliasConfig, svc *templat
...
+	// changes that can be applied in the running server
+	if oldEP.IP != newEP.IP || oldEP.Port != newEP.Port {
+		if err := b.innerUpdateServerAddrPort(newEP); err != nil {
+			return false, err
+		}
+	}
+
+	if oldEP.Weight != newEP.Weight {
+		return false, b.innerUpdateServerWeight(newEP, isPassthrough)
+	}
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/configmanager/haproxy/backend.go:363]: Good catch, addressed by adding a new method to properly orchestrate changes in weight.

File: pkg/router/template/configmanager/haproxy/backend.go [hunk 2]
```diff
@@ -331,33 +343,63 @@ func (b *Backend) AddServer(cfg *templaterouter.ServiceAliasConfig, svc *templat
...
+}
+
+// ReplaceServer dynamically replaces the backend server by removing it and adding again with new configuration.
+func (b *Backend) ReplaceServer(cfg *templaterouter.ServiceAliasConfig, svc *templaterouter.ServiceUnit, oldEP, newEP *templaterouter.Endpoint, workingDir, defaultDestinationCA string) error {
+	if err := b.innerSetServerState(oldEP, false); err != nil {
+		return err
+	}
+	if err := b.innerDeleteServer(oldEP); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/configmanager/haproxy/backend.go:395]: Adding or enabling server should also fail, but in fact fail to delete is more likely to happen due to keepalive / in flight requests. So, both added your suggestion and improved godoc of the method.

File: pkg/router/template/configmanager/haproxy/manager.go [hunk 3]
```diff
@@ -447,75 +447,102 @@ func (cm *haproxyConfigManager) ReplaceRouteEndpoints(id templaterouter.ServiceA
...
+					if err == nil {
+						// Server replaced successfully, mark to enable health check later.
+						newEPs = append(newEPs, ep)
+					}
+				}
+				if err != nil {
+					// Failed either enabling health check or replacing backend server.
+					errs = append(errs, err)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/configmanager/haproxy/manager.go:551]: I see your point. I however used a distinct approach, let me know if I'm missing anything the way I did.

File: pkg/router/template/router.go [hunk 4]
```diff
@@ -855,14 +852,56 @@ func (r *templateRouter) dynamicallyAddRoute(backendKey ServiceAliasConfigKey, r
...
+				failed = true
+			}
+		}
+	}
+
+	for serviceKey := range oldConfig.ServiceUnits {
+		if _, newFound := newConfig.ServiceUnits[serviceKey]; !newFound {
+			if service, ok := r.findMatchingServiceUnit(serviceKey); ok {
+				// Service was removed from route, remove the endpoints that were in the old config.
+				if err := r.dynamicConfigManager.RemoveRouteEndpoints(backendKey, service.EndpointTable); err != nil {
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/router.go]: Addressed, thank you!

**Comment by jcmoraisjr** [comment]
> or even have a separate revert commit if it reverts cleanly.

Added a separate commit reverting https://github.com/openshift/router/commit/0e47c836e9f2e5bbbf95c35604fc89a6fa8b32ee, the new implementation is on top of it.

File: pkg/router/template/configmanager/haproxy/manager.go [hunk 5]
```diff
@@ -737,6 +763,12 @@ func (cm *haproxyConfigManager) reset() {
 // findMatchingBlueprint finds a matching blueprint route that can be used
 // as a "surrogate" for the route.
 func (cm *haproxyConfigManager) findMatchingBlueprint(route *routev1.Route) *routev1.Route {
+
+	// HAProxy 2.8 is not working well adding backend servers on an empty backend, like blueprint servers.
+	// Blueprint servers are being removed via https://redhat.atlassian.net/browse/NE-2663, so we're
+	// just anticipating its deprecation by not using it in case it is being configured.
+	return nil
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/configmanager/haproxy/manager.go:775]: This is an intended skip for simplicity: haproxy 2.8 branch has some issues adding servers dynamically on backends without any server, and blueprints are going to be completely removed on a followup PR. We're going to cut the blueprint path here, and the rest of the code is going to be removed via https://redhat.atlassian.net/browse/NE-2663

File: pkg/router/template/configmanager/haproxy/manager.go [hunk 6]
```diff
@@ -447,75 +447,103 @@ func (cm *haproxyConfigManager) ReplaceRouteEndpoints(id templaterouter.ServiceA
...
+		newEP := newEndpoints[i]
+		j := slices.IndexFunc(oldEndpoints, func(oldEP templaterouter.Endpoint) bool {
+			return oldEP.ID == newEP.ID
 		})
-		if existing {
-			modifiedEndpoints[ep.ID] = ep
+		if j >= 0 {
+			oldEP := oldEndpoints[j]
+			if !reflect.DeepEqual(oldEP, newEP) {
+				modifiedEndpoints[newEP.ID] = epPair{oldEP: &oldEP, newEP: &newEP}
```
[the remark this answers is not in the document]
Comment (jcmoraisjr) [reply pkg/router/template/configmanager/haproxy/manager.go:466]: Just addressed this one by skipping dynamic update in the case of a change in NoHealthCheck for now. This would need a non trivial update, and the change that triggers this update is a non recurrent one. We can revisit this in case this becomes a recurrent type of update.
